### PR TITLE
Optimize terrain

### PR DESCRIPTION
**Pull request by akortunov**
Task [#4695](https://gitlab.com/OpenMW/openmw/issues/4695).

In my testing OpenMW now consumes 2GB of memory on start with the Tamriel Rebuilt and Stirk instead of 5.5Gb.

Also helps much with freezes during turning.
Freezes still may happen, but only with insane viewing distance (there are some fixes in bzzt's brach which can help here).

The only issue is default cells handling. Usually we render some default cells near player and it is enough when the Distant Terrain is disabled, otherwise player can see holes in terrain. Water shader allows to mitigate this issue.

What we should do here? Assigning default data for every cells (as we do in the 0.45) is too expensive). Suggest users to enable water shader with Distant Terrain?

### CODE REVIEW

**Comment by AnyOldName3** [comment]
One idea I've discussed in the past would be to just reuse the same terrain drawable for everywhere with default terrain, just with a different transform matrix.

File: apps/openmw/mwrender/renderingmanager.cpp [hunk 1]
```diff
@@ -1195,6 +1195,10 @@ namespace MWRender
 
         mUniformNear->set(mNearClip);
         mUniformFar->set(mViewDistance);
+
+        // Since our fog is not radial yet, we should take FOV in account, otherwise terrain near viewing distance may disappear.
+        float distanceMult = std::cos(osg::DegreesToRadians(mFieldOfView)/2.f);
+        mTerrain->setViewDistance(mViewDistance * (distanceMult ? 1.f/distanceMult : 1.f));
```
Comment (Capostrophic) [inline apps/openmw/mwrender/renderingmanager.cpp]: As something I haven't thought about, on extremely high FOVs (beyond 180 degrees) the mult becomes negative and bad things happen. Either distant mult should be absolute, or be replaced with a cosine of (pi-mFieldOfView/2.f) before usage, or there simply should be a check that distance mult is positive.

Comment (wareya) [reply apps/openmw/mwrender/renderingmanager.cpp]: me: "that doesn't work though!"

me: test it out ingame

me: https://i.imgur.com/HJooPAl.jpg

Comment (Capostrophic) [reply apps/openmw/mwrender/renderingmanager.cpp]: Yeah. The problem is that with these changes the terrain disappears too. My bad, I wrote this part.
screenshot021

**Comment by unelsson** [comment]
Tested this with two huge landmasses, and also Morrowind, all with view distance 350000, compared this to master, and the improvement is great. Framerates with huge landmasses are 250-450, whereas master branch crashes (OOM with 24gb ram). Ram usage with such settings goes somewhere between 3-4gb (in system setup with 2,1gb RAM used before running OpenMW, that is OpenMW adds another uses 1-2gb ram with view distance 350k). Flew around with speeds 50, 500, 5000 and 50k. Looked around landmass shadow rendering, compared it to master, and can't see any problems with that. Asked to test with high FOV: Indeed beyond 180 degrees it goes haywire, but FOV slightly less than 180 does work, although obviously doesn't look very nice on a normal screen.

**Comment by AnyOldName3** [comment]
How are you setting the fov above 180? Projection matrices don't make sense at or above 180.

**Comment by wareya** [comment]
editing the config manually

**Comment by AnyOldName3** [comment]
If I stick -50 as the full screen resolution, I can put that in the config, but that doesn't mean it's going to work.

**Comment by wareya** [comment]
yes, obviously

**Comment by psi29a** [comment]
Do we need to put a min/max bound on FOV to prevent craziness? And also warn user about non-sane config value?

**Comment by AnyOldName3** [comment]
I'd just assumed something would be spitting out an error message at some point whether it was us upon encountering the value or OSG upon being asked to use the value. If no error is printed, we should probably make it be.

I don't think it's reasonable to bother making the distant terrain system robust against FOVs `>=` 180°, as that's never going to be achievable by a single camera, but maybe we should be making considerations for things like 360° screenshots where several different viewing angles are used back to back.

File: apps/openmw/mwrender/renderingmanager.cpp (hunk 1 above)
Comment (akortunov) [reply apps/openmw/mwrender/renderingmanager.cpp]: And which cap we should use for FOV? 110, as in UI?

Comment (wareya) [reply apps/openmw/mwrender/renderingmanager.cpp]: please no, there are valid reasons to play with unreasonably high FoVs, making the config file cap 179 or something just to keep it from being a broken value is fine

should also enforce a minimum of like 1 or something if we don't already

**Comment by akortunov** [comment]
Did anyone test this PR with 360° screenshots?

**Comment by Capostrophic** [comment]
https://vizor.io/capostrophic/test/

So, yeah, I don't think it's fair to rely on refraction to hide the missing seafloor. I think that one cell (there are more but they are loaded around the player) is missing from Tamriel Rebuilt and from Morrowind.esm and Firemoth should be there.

**Comment by akortunov** [comment]
So what we decided to do here?

**Comment by Capostrophic** [comment]
Not having something universal to substitute default terrain without ridiculous memory usage is pretty sad and I still think it's desirable to look into a way to add that. But I think everyone can agree the terrain viewing distance limit thing could probably be merged straightaway.

**Comment by akortunov** [comment]
> Not having something universal to substitute default terrain without ridiculous memory usage is pretty sad and I still think it's desirable to look into a way to add that.

Even in upstream master, the landscape plane is not endless anyway - we are limited by the QuadTree size, and the border is visible in places such as Solstheim. I see no reliable way to generate endless terrain plain - transparent vanilla water just was not designed for large viewing distance.

I tried to theck how MGE handles default cells, but it seems there is no easy way to disable water shader in MGE (even `togglewater` does not work).

Can we somehow limit viewing distance for things behind the non-shader water?

**Comment by Capostrophic** [comment]
Won't argue with the first bunch of statements, but if I remember correctly this drops *any* cells that don't have any data and not just those "out of the bounds of the actual terrain" like it happened in 0.43.0 and earlier which leads to ugly holes like on the screenshot above. While 0.43.0 method obviously would make memory usage higher, surely it wasn't that bad if the ram usage only went bonkers once the entire quad tree was loaded into the memory. Couldn't it and the new approach (load cells around the player) be combined somehow?

**Comment by akortunov** [comment]
> but if I remember correctly this drops any cells that don't have any data and not just those "out of the bounds of the actual terrain" like it happened in 0.43.0

Nope. This PR uses a same approach to render default cells, as our common terrain system - renders grid near player. 
In 0.43 we indeed did not render cells without terrain at all with Distant Terrain.

**Comment by lysol90** [comment]
> Can we somehow limit viewing distance for things behind the water? We already do it, but only when the player is underwater.

Yes please. This would make a big difference I think. In screenshots taken from high above, you see right down to the ocean floor as long as the water isn't in an oblique angle. Limiting the view distance to prevent this would look much better. If that is possible at the moment of course.

**Comment by akortunov** [comment]
@AnyOldName3, can we implement it without shaders?

**Comment by AnyOldName3** [comment]
If you do it with fog, the fixed function pipeline can do it. I'm not sure how exactly we can make that work, though. If refraction is on, we can set a different fog range for its RTT, but otherwise, I don't think there's any way to tell the FFP to use one fog distance for things below z=0 in world coordinates and another for things above z=0.

**Comment by akortunov** [comment]
> I don't think there's any way to tell the FFP to use one fog distance for things below z=0 in world coordinates and another for things above z=0.

So it seems that the only way to make water with Distant Terrain to look good is to use refraction?

File: apps/openmw/mwrender/renderingmanager.cpp [hunk 2]
```diff
@@ -1195,6 +1197,10 @@ namespace MWRender
...
```
Comment (Capostrophic) [inline apps/openmw/mwrender/renderingmanager.cpp]: I think it should still make sense to put an upper limit on the multiplier, something like 2 (110 degrees FOV) or 3 (140 degrees FOV). Or it'll be 114 if 179 degree FOV is used.

**Comment by tcotys** [comment]
I might be a bit off topic but there is lenses that uses a FOV > 180°.
It depends on the type of the projection used for the camera :
- the usual orthographic FOV projection is limited at 180°
- but (for exemple) stereographic projection can get up to FOV = 360° (i.e. the "small planet" effet)
with a relatively small FOV both projections a quite the same

**Comment by AnyOldName3** [comment]
That doesn't mean that the maths of projection matrices works like that, though. A projection matrix converts a frustum to a cube, and frustums don't make sense as a concept with an angle at or above 180°. When rendering things to look like they've got a FOV above 180°, it has to be done by rendering multiple times with a smaller FOV, facing in different directions, and combining the results together.

File: apps/openmw/mwrender/renderingmanager.cpp [hunk 3]
```diff
@@ -1200,6 +1202,8 @@ namespace MWRender
         mUniformFar->set(mViewDistance);
 
         // Since our fog is not radial yet, we should take FOV in account, otherwise terrain near viewing distance may disappear.
+        // Limit FOV here just for sure, otherwise viewing distance can be too high.
+        fov = std::min(fov, 140.f);
         float distanceMult = std::cos(osg::DegreesToRadians(mFieldOfView)/2.f);
```
Comment (Capostrophic) [inline apps/openmw/mwrender/renderingmanager.cpp]: fov variable is not used.

**Comment by akortunov** [comment]
I suggest to recommend users to use a refraction with Distant Terrain for now and fill the feature request about default cells handling.
But as I said, I see no reliable to limit the underwater viewing distance with the fixed function pipeline.

Any other objections about this PR?

**Comment by psi29a** [comment]
Last call...

**Comment by AnyOldName3** [comment]
I'm a little confused why we're using the terrain grid to render default cells instead of doing it with the same system as the rest of distant terrain.

**Comment by akortunov** [comment]
> I'm a little confused why we're using the terrain grid to render default cells instead of doing it with the same system as the rest of distant terrain.

We can not use the same system here - the quad tree size is limited.
Ideally, we need to use a some kind of paging here to cover the whole visible range for both terrain and water (yes, our water plane is not infinite too), but it is outside of scope of this PR.

**Comment by Capostrophic** [comment]
>yes, our water plane is not infinite too

It's not, but its origin moves with the player, which makes the fact only noticeable on a humongous viewing distance.

**Comment by akortunov** [comment]
So what's up with this PR?

**Pull request merged**